File: ios/thunderid_flutter/Sources/thunderid_flutter/ManagementMethodHandler.swift
// Copyright 2026 The ThunderID Authors
// SPDX-License-Identifier: Apache-2.0

import Foundation
import ThunderID

/// Routes `applications.*`, `users.*`, and `agents.*` method channel calls to the native
/// management API. Resources cross the channel as JSON maps, encoded and decoded with the
/// native models' own `Codable` conformance.
@MainActor
enum ManagementMethodHandler {
    static func handles(_ method: String) -> Bool {
        ["applications", "users", "agents"].contains(method.split(separator: ".").first.map(String.init) ?? "")
    }

    static func handle(method: String, args: [String: Any], client: ThunderIDClient) async throws -> Any? {
        let id = args["id"] as? String ?? ""
        let limit = args["limit"] as? Int
        let offset = args["offset"] as? Int
        switch method {
        case "applications.list":
            return try encode(try await client.applications.list(limit: limit, offset: offset))
        case "applications.get":
            return try encode(try await client.applications.get(id: id))
        case "applications.create":
            return try encode(try await client.applications.create(decode(args["payload"])))
        case "applications.update":
            return try encode(try await client.applications.update(id: id, decode(args["payload"])))
        case "applications.delete":
            try await client.applications.delete(id: id)
            return nil
        default:
            return try await handleUsersAndAgents(method: method, args: args, client: client)
        }
    }

    private static func handleUsersAndAgents(
        method: String,
        args: [String: Any],
        client: ThunderIDClient
    ) async throws -> Any? {
        let id = args["id"] as? String ?? ""
        let limit = args["limit"] as? Int
        let offset = args["offset"] as? Int
        switch method {
        case "users.list":
            let filter = args["filter"] as? String
            return try encode(try await client.users.list(limit: limit, offset: offset, filter: filter))
        case "users.get":
            return try encode(try await client.users.get(id: id))
        case "users.create":
            return try encode(try await client.users.create(decode(args["payload"])))
        case "users.update":
            return try encode(try await client.users.update(id: id, decode(args["payload"])))
        case "users.delete":
            try await client.users.delete(id: id)
            return nil
        case "agents.list":
            return try encode(try await client.agents.list(limit: limit, offset: offset))
        case "agents.get":
            return try encode(try await client.agents.get(id: id))
        case "agents.create":
            return try encode(try await client.agents.create(decode(args["payload"])))
        case "agents.update":
            return try encode(try await client.agents.update(id: id, decode(args["payload"])))
        case "agents.delete":
            try await client.agents.delete(id: id)
            return nil
        default:
            throw ThunderIDError(code: .unknownError, message: "Unknown method: \(method)")
        }
    }

    private static func encode<T: Encodable>(_ value: T) throws -> Any {
        try JSONSerialization.jsonObject(with: JSONEncoder().encode(value))
    }

    private static func decode<T: Decodable>(_ value: Any?) throws -> T {
        guard let value, JSONSerialization.isValidJSONObject(value) else {
            throw ThunderIDError(code: .invalidInput, message: "A payload is required")
        }
        return try JSONDecoder().decode(T.self, from: JSONSerialization.data(withJSONObject: value))
    }
}

File: android/src/main/kotlin/dev/thunderid/flutter/ManagementMethodHandler.kt
// Copyright 2026 The ThunderID Authors
// SPDX-License-Identifier: Apache-2.0

package dev.thunderid.flutter

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import dev.thunderid.android.IAMException
import dev.thunderid.android.ThunderIDClient
import dev.thunderid.android.ThunderIDErrorCode
import dev.thunderid.android.management.ApplicationRequest
import dev.thunderid.android.management.CreateAgentRequest
import dev.thunderid.android.management.CreateManagedUserRequest
import dev.thunderid.android.management.UpdateAgentRequest
import dev.thunderid.android.management.UpdateManagedUserRequest

/**
 * Routes `applications.*`, `users.*`, and `agents.*` method channel calls to the native
 * management API. Resources cross the channel as JSON maps, encoded and decoded with Gson,
 * which the native SDK already uses for its models.
 */
internal object ManagementMethodHandler {
    private val gson = Gson()
    private val mapType = object : TypeToken<Map<String, Any?>>() {}.type

    fun handles(method: String): Boolean = method.substringBefore('.') in setOf("applications", "users", "agents")

    suspend fun handle(method: String, args: Map<String, Any?>, client: ThunderIDClient): Any? {
        val id = args["id"] as? String ?: ""
        val limit = (args["limit"] as? Number)?.toInt()
        val offset = (args["offset"] as? Number)?.toInt()
        return when (method) {
            "applications.list" -> encode(client.applications.list(limit, offset))
            "applications.get" -> encode(client.applications.get(id))
            "applications.create" -> encode(client.applications.create(decode<ApplicationRequest>(args)))
            "applications.update" -> encode(client.applications.update(id, decode<ApplicationRequest>(args)))
            "applications.delete" -> client.applications.delete(id).let { null }
            "users.list" -> encode(client.users.list(limit, offset, args["filter"] as? String))
            "users.get" -> encode(client.users.get(id))
            "users.create" -> encode(client.users.create(decode<CreateManagedUserRequest>(args)))
            "users.update" -> encode(client.users.update(id, decode<UpdateManagedUserRequest>(args)))
            "users.delete" -> client.users.delete(id).let { null }
            "agents.list" -> encode(client.agents.list(limit, offset))
            "agents.get" -> encode(client.agents.get(id))
            "agents.create" -> encode(client.agents.create(decode<CreateAgentRequest>(args)))
            "agents.update" -> encode(client.agents.update(id, decode<UpdateAgentRequest>(args)))
            "agents.delete" -> client.agents.delete(id).let { null }
            else -> throw IAMException(ThunderIDErrorCode.UNKNOWN_ERROR, "Unknown method: $method")
        }
    }

    private fun encode(value: Any): Map<String, Any?> = gson.fromJson(gson.toJson(value), mapType)

    private inline fun <reified T> decode(args: Map<String, Any?>): T {
        val payload = args["payload"] ?: throw IAMException(ThunderIDErrorCode.INVALID_INPUT, "A payload is required")
        return gson.fromJson(gson.toJson(payload), T::class.java)
    }
}
